R/estimate_KM.R

```diff
@@ -65,102 +65,56 @@
- ### df: catch expressions that represent base R subsets
-  Call <- as.list(match.call())
-  dfExpr <- Call[["data"]]
-
- ## Validate `data` and capture data name
-
-  if (is.null(data)) stop(paste0("Data can't be NULL."))
-
-  if (base::length(base::deparse(Call[["data"]])) == 1 && base::deparse(Call[["data"]]) %in% c(".", ".x")){
-    df <- the_lhs()
-    Call[["data"]] <- as.symbol(df)
-  } else {
-     df <- as.character(sub("\\[.*$", "", deparse(dfExpr))[1])
-  }
-
-  if (!(inherits(data, "data.frame") | inherits(data, "tibble") | inherits(data, "data.table"))) stop(paste0("Data can be of `class` dataframe or tibble or data.table."))
-
-  data <- as.data.frame(data)
-
-# Validate columns --------------------------------------------------------
+  ## Validate argument inputs --------------------------------------------------
+  if (is.null(data))
+    stop(paste0("Data can't be NULL."))
+  if (!is.data.frame(data))
+    stop("Data can be of `class` dataframe or tibble or data.table.")
 
   reqcols <- c(strata, CNSR, AVAL)
-
   if (! all(reqcols %in% colnames(data))){
     stop(paste0("Following columns are missing from `data`: ", paste(setdiff(reqcols, colnames(data)), collapse = " "), "."))
   }
 
-  if (! is.numeric(data[[AVAL]])){
+  if (!is.numeric(data[[AVAL]])){
     stop("Analysis variable (AVAL) is not numeric.")
   }
 
-  if (! is.numeric(data[[CNSR]])){
+  if (!is.numeric(data[[CNSR]])){
     stop("Censor variable (CNSR) is not numeric.")
   }
 
-# Ensure the presence of at least one strata -----------------------------
-
-  if (is.null(strata)){
-    main <- "1"
-  } else {
-    main <- paste(strata, collapse = " + ")
-  }
-
-# Remove NA from the analysis --------------------------------------------
-
-  data <- as.data.frame(data) %>%
-    tidyr::drop_na(AVAL, CNSR)
-
-  if (!is.null(strata)){
-    data <- data %>%
-      tidyr::drop_na(any_of({{strata}}))
-  }
+  # Remove NA from the analysis ------------------------------------------------
+  data <-
+    as.data.frame(data) %>%
+    tidyr::drop_na(any_of(c(AVAL, CNSR, strata)))
 
-# Calculate survival and add time = 0 to survfit object -------------------
+  # construct survfit() call ---------------------------------------------------
+  formula_rhs <-
+    ifelse(is.null(strata), "1", paste(strata, collapse = " + "))
+  formula <- stats::as.formula(paste0("survival::Surv(", AVAL, ", 1-", CNSR, ") ~ ", formula_rhs))
 
- ## Reverse censoring: see ADaM guidelines versus R survival KM analysis
+  survfit_object <-
+    rlang::inject(survival::survfit(!!formula, data = data, !!!dots)) %>%
+    survival::survfit0(start.time = 0)
 
-  formula <- stats::as.formula(paste0("survival::Surv(", AVAL, ", 1-", CNSR, ") ~ ", main))
+  # convert survfit() call to quo with attached envir --------------------------
+  survfit_object$call[[1]] <- rlang::expr(survival::survfit) # adding `survival::` prefix
+  survfit_object$call <- rlang::quo(!!survfit_object$call)
```

Comment:
For me two issues here:

1.  sf <- adtte %>%
  dplyr::filter(TRTA == "Placebo") %>%
  estimate_KM(strata = "SEX")

  > sf
Call: ~survival::survfit(formula = survival::Surv(AVAL, 1 - CNSR) ~ 
    SEX, **data = data**)
I want it to say data = "sf"

2. With this change we are breaking the compatibility with traditional survfit objects. This we wish to avoid. Could we save the environment alone into an argument and then reconstruct the call if an environment is attached? Otherwise we redo the call as we did before?

Reply:
1. The sf name is saved in `data_name`, not in the quo. The quo must evaluate properly and this would not happen if we changed the data object.
2. I started looking into this, and was happy to see that everything works wit the `survfit()` object as well. There was only once instance I tested that errored and that is when the survfit object is created like `data %>% filter() %>% survfit(one ~ two, data = .)`, `get_pvalue()` is later called on the object. BUT this has never worked in the package, and is a key example of when you need the correct envir specified to evaluated.

Reply:
The main issue is that I wish to see the correct call when printing the object. data = data is just not informative. Since the mofied dataframe is inside the same environment, I don't see why it can't be captured. Perhaps we should create a new variable with the name of the actual data 'name of original df' <- data and then create the quosure. Would that work?

Reply:
The main issue is that I wish to see the correct call when printing the object. data = data is just not informative. Since the mofied dataframe is inside the same environment, I don't see why it can't be captured. Perhaps we should create a new variable with the name of the actual data 'name of original df' <- data and then create the quosure. Would that work?

Reply:
For the second issue, we need to investigate further. Did you find why it does not work?

Reply:
1. expressions, quosures, and environments are already difficult to handle. The more we massage them and manipulate them, we are likely to not cover edge cases and will introduce bugs. Also, we don't have a reliable method for extracting the data set symbol, which is why we are here adding quosures and environments: our previous method led to a silent bug when trying to extract the input data set symbol and use it in `get_pvalue()`. The tilda before the quosure is how the are printed and cannot be modified. The raw output from a survfit object isn't useful anyway, so let's not complicate the situation by trying to modifying it.
2. I do know why this doesn't work. If you construct a `survfit()` using `%>%` and `data = .`, the saved call will contain `data = .`. To calculate the p-value, the call is parsed and reconstructed into a `survdiff()` call, but `data = .` can never be recovered. We could improve error messaging in this situation, but there is nothing we can do to avoid the error.